nixos/modules/system/boot/systemd/initrd.nix

```diff
@@ -0,0 +1,366 @@
+
+  kernel-name = config.boot.kernelPackages.kernel.name or "kernel";
+  modulesTree = config.system.modulesTree.override { name = kernel-name + "-modules"; };
+  firmware = config.hardware.firmware;
+  # Determine the set of modules that we need to mount the root FS.
+  modulesClosure = pkgs.makeModulesClosure {
+    rootModules = config.boot.initrd.availableKernelModules ++ config.boot.initrd.kernelModules;
+    kernel = modulesTree;
+    firmware = firmware;
+    allowMissing = false;
+  };
+
+  emergencyEnv = pkgs.buildEnv {
+    name = "initrd-emergency-env";
+    paths = map getBin cfg.emergencyPackages;
+    pathsToLink = ["/bin" "/sbin"];
+  };
+
+  initialRamdisk = pkgs.makeInitrdNG {
+    contents = cfg.objects;
+  };
+
+in {
+  options.boot.initrd.systemd = {
+    enable = mkEnableOption ''
+      Use systemd in initrd.
+
+      Note: This is in very early development and is highly
+      experimental. Most of the features NixOS supports in initrd are
+      not yet supported by the intrd generated with this option.
+    '';
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.systemdMinimal;
+      defaultText = "pkgs.systemdMinimal";
+      description = ''
+        The systemd package to use for initrd.
+      '';
+    };
+
+    objects = mkOption {
+      description = "List of objects to include in the initrd, and their symlinks";
+      example = ''
+        [ { object = "''${systemd}/lib/systemd/systemd"; symlink = "/init"; } ]
+      '';
+      type = types.listOf (types.submodule {
+        options = {
+          object = mkOption {
+            type = types.path;
+            description = "The object to include in initrd.";
+          };
+          symlink = mkOption {
+            type = types.nullOr types.path;
+            description = "A symlink to create in initrd pointing to the object.";
+            default = null;
+          };
+        };
+      });
+    };
+
+    emergencyHashedPassword = mkOption {
+      type = types.str;
+      description = ''
+        Hashed password for the super user account in stage 1 emergency mode
+
+        Blank for no password, ! for super user disabled.
+      '';
+      default = "!";
+    };
+    emergencyPackages = mkOption {
+      type = types.listOf types.package;
+      default = [];
+      description = ''
+        Packages to include in /bin for the stage 1 emergency shell.
+      '';
+    };
+
+    additionalUpstreamUnits = mkOption {
+      default = [ ];
+      type = types.listOf types.str;
+      example = [ "debug-shell.service" "systemd-quotacheck.service" ];
+      description = ''
+        Additional units shipped with systemd that shall be enabled.
+      '';
+    };
+    suppressedUnits = mkOption {
+      default = [ ];
+      type = types.listOf types.str;
+      example = [ "systemd-backlight@.service" ];
+      description = ''
+        A list of units to suppress when generating system systemd configuration directory. This has
```

Comment:
What does "suppress a unit" mean in this context? I find that unclear unless I look at the implementation

Reply:
What would you suggest this say?